### DISCUSSION TITLE
Performance differences across browsers

### DISCUSSION THREAD

**Discussion by oaklandgit** [replit/kaboom#365] [Q&A] [answered]
Hi,

My game runs great (60 fps) on Chrome for Mac OS, but terribly on Safari and Firefox (< 20 fps).

On Safari, it could have something to do with the fact that Safari is not respecting the "crisp: true" setting, and is rendering my game as anti-aliased.

In Firefox, the game renders crisp, but it's still slow.

https://tailpipe-10-24.larrystone.repl.co/

Any ideas? Or perhaps there's something in the Kaboom rendering code that is browser-specific?

**Comment by oaklandgit** [comment]
Update:

I've experimented with different approaches to culling offscreen objects. Nothing seems to help in getting Mac OS Safari above 20fps. I've been able to improve Firefox to about 30fps. Meanwhile, Chrome runs beautifully.

**Comment by slmjkdbtl** [comment]
I'll do some profiling and get back to you if identified the source

**Comment by oaklandgit** [comment]
More info:

I found a performance hog by profiling with Chrome. I then removed that hog and saw that it improved Safari somewhat. Then I noticed this:

Safari maxes out at 30fps even when I'm in my title scene (with literally nothing happening except displaying text and waiting for a keypress.)

![image](https://user-images.githubusercontent.com/421615/140461075-fd6d660a-a58e-428c-937b-a6dcc194d822.png)

So in short, it seems that Safari is limiting framerate out of the box, so to speak.

**Reply by slmjkdbtl** [reply]
Weird, what about stuff on https://kaboomjs.com/play ? Do those also have 30fps on safari?

**Reply by oaklandgit** [reply]
Great question, and indeed it does:
![image](https://user-images.githubusercontent.com/421615/140462838-d70f8f01-f62d-4e7f-b8b2-526fd2223440.png)

**Reply by oaklandgit** [reply]
Ahh. I googled it and see that it could be a low-power mode thing. Stay tuned.

**Reply by oaklandgit** [reply]
Well, that there's my problem. User error.

Mac has a low power mode that throttles Safari framerate. I turned it off and zooom. Off to the races!

**Comment by oaklandgit** [comment, marked as answer]
Solved: Safari throttles framerate when Mac OS is in Low Power Mode. Turn it off for maximum Kaboom.

![image](https://user-images.githubusercontent.com/421615/140463332-ed6747b1-34c0-421c-8ac6-90aba7a589ff.png)
